leakcanary-deobfuscation-gradle-plugin/src/main/java/com/squareup/leakcanary/deobfuscation/LeakcanaryLeakDeobfuscationPlugin.kt

```diff
@@ -0,0 +1,111 @@
+package com.squareup.leakcanary.deobfuscation
+
+import com.android.build.gradle.AppExtension
+import com.android.build.gradle.BaseExtension
+import com.android.build.gradle.LibraryExtension
+import com.android.build.gradle.api.ApplicationVariant
+import com.android.build.gradle.api.BaseVariant
+import com.android.build.gradle.api.LibraryVariant
+import org.gradle.api.DefaultTask
+import org.gradle.api.DomainObjectSet
+import org.gradle.api.GradleException
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+import org.gradle.api.Task
+import org.gradle.api.UnknownTaskException
+import org.gradle.api.tasks.TaskProvider
+import java.io.File
+
+class LeakcanaryLeakDeobfuscationPlugin : Plugin<Project> {
+  override fun apply(project: Project) {
+    val androidPluginHandler = { _: Plugin<*> ->
+      val extension = project.extensions.getByType(BaseExtension::class.java)
+      val variants: DomainObjectSet<BaseVariant> = when (extension) {
+        is AppExtension -> {
+          extension.applicationVariants as DomainObjectSet<BaseVariant>
+        }
+        is LibraryExtension -> {
+          extension.libraryVariants as DomainObjectSet<BaseVariant>
+        }
+        else -> throw GradleException(
+            "This plugin can only be applied in Android application or library modules."
+        )
+      }
+      val leakCanaryExtension = project.extensions.create(
+          "leakCanary",
+          LeakCanaryDeobfuscationExtension::class.java
+      )
+
+      variants.all { variant ->
+        val obfuscatedVariant =
+          leakCanaryExtension.filterObfuscatedVariants?.invoke(variant) ?: false
+        if (obfuscatedVariant || variant.buildType.isMinifyEnabled) {
```

Comment:
you should not have a second condition here. We want devs to be able to fully control filtering. 
Sorry if I was not clear, but `variant.buildType.isMinifyEnabled` should not be here: it should be the default value of `leakCanaryExtension.filterObfuscatedVariants` . Devs will override it if they want using the DSL. It will never be null.

Reply:
But the extension doesn't know about `variant`. I played with passing `variant` or this default value to an extension but each field in an extension is a public api of the plugin so I wasn't sure if we should expose that.

Reply:
I don't understand: in the extension, do: `filterObfuscatedVariants = {variant: BaseVariant -> variant.buildType.isMinifyEnabled)} `

Reply:
Ah 🤦‍♂ We have it because users pass this variant from their build scripts! Sorry I completely misunderstood this, but it's clear now. And thank you for your great comments and bearing with me 😉

Reply:
@stephanenicolas 
I was thinking about it more and I realized that having `variant.buildType.isMinifyEnabled` as a default value may not be a good idea.
If user adds this plugin to his project but forgets about adding configuration (`leakCanary` block) then all variants with minification enabled will have mapping file copied to assets directory. And this is not good from the security point of view. I thi k `false` is a safer default. It's better to not copy the mapping than copying it by mistake.
What do you think?

Reply:
You're totally right, or we could precise to include only variant whose buildtype name is debug and that are minified. Up to you, or false.

Reply:
I'm also thinking that this plugin should be used specifically for LeakCanary purpose. Maybe we could somehow check if LeakCanary is present in the project and if it's not then skip executing the task or even crash. Do you think it's a good idea? And if it is then do you know how could we check if LC is present?